log bytes recv/sent per command

jonasschnelli opened the pull request:
Not sure if we want to have this is master.

This adds a sent/recv byte counter per command to each `CNode`.
It is a waste product of some statistics gathering during finding a strategy for throttling nodes without affecting the overall quality of the network.

I think throttling the `block` response to a node with a height < self.height - 288 (TBD) would be a good approach. Nodes in initial sync might need longer to catch up if they are connected to throttled nodes only. Initial sync is a one time process and is therefore (maybe) be not very time critical.

Example:

```
    "bytessent_per_cmd": {
      "addr": 30003,
      "block": 28981980,
      "getheaders": 9682,
      "headers": 82,
      "inv": 30843,
      "ping": 48,
      "pong": 40,
      "tx": 97400,
      "verack": 0,
      "version": 103
    },
    "bytesrecv_per_cmd": {
      "getaddr": 0,
      "getdata": 11935,
      "getheaders": 997,
      "headers": 1458027,
      "ping": 40,
      "pong": 40,
      "reject": 230,
      "verack": 0,
      "version": 102
    }
```

jonasschnelli committed:
log bytes recv/sent per command

sipa commented on src/net.cpp:
```diff
@@ -2171,7 +2174,7 @@ void CNode::AbortMessage() UNLOCK_FUNCTION(cs_vSend)
 }
 
-void CNode::EndMessage() UNLOCK_FUNCTION(cs_vSend)
+void CNode::EndMessage(const char* pszCommand) UNLOCK_FUNCTION(cs_vSend)
```
Any reason not to use const std::string& ?

jonasschnelli replied:
`CNode::BeginMessage()` also uses `const char*` (https://github.com/bitcoin/bitcoin/blob/master/src/net.h#L499). `EndMessage()` is always called over `PushMessage(const char* pszCommand, const T1& a1, ...)`. Not sure if it would make sense to cast/transform to a `std::string`.

sipa replied:
ok

laanwj replied:
Right, I agree with @sipa that I prefer std::string, but yes for that to have any effect whatsoever that'd require switching that entire call chain to `std::string` and that is out of the scope of this pull.

paveljanik replied:
On the other hand, we are already changing `EndMessage()` here, os it could be done in one step, together.

jonasschnelli replied:
I think it's to invasive for a little logging feature. It would basically require to change all `PushMessage(...)` then `BeginMessage()` and `EndMessage()` function. An either conversion of CMessageHeader (that it accepts `std::string`) or some conversions to c-strings in the mentioned functions.
IMO, it's not worth changing this.

paveljanik replied:
Agreed.

sipa commented on src/net.cpp:
```diff
@@ -613,7 +613,9 @@ void CNode::copyStats(CNodeStats &stats)
     X(fInbound);
     X(nStartingHeight);
     X(nSendBytes);
+    X(mSendBytesPerCmd);
```
I don't care about following the original variable naming style, but if you do, the prefix for maps is map* not m*.

sipa commented on src/net.cpp:
```diff
@@ -2194,6 +2197,9 @@ void CNode::EndMessage() UNLOCK_FUNCTION(cs_vSend)
     WriteLE32((uint8_t*)&ssSend[CMessageHeader::MESSAGE_SIZE_OFFSET], nSize);
 
+    //log total amount of bytes per command
+    mSendBytesPerCmd[std::string(pszCommand)]+=nSize;
```
Spaces around the operator?

sipa commented:
Concept ACK, I think this is useful.

jonasschnelli commented:
Thanks @sipa for reviewing!
Fixed nits. Added `bytessent_per_cmd` and `bytesrecv_per_cmd` to `getpeersinfo` help.

sipa commented:
I haven't checked the code for this, but we should make sure this only
accounts for incoming messages of a known type. Otherwise you introduce a
memory DoS attack by allowing the sender to construct an arbitrarily-sized
map.

jonasschnelli commented:
@sipa: good point! Will address it soon.

theuni commented:
These maps will need locking

jonasschnelli commented:
@theuni: Right. I was aware of the locking and did come to the conclusion that both new maps are protected by `cs_vSend` and `cs_vRecvMsg`.

Moved the maps to the private section.

jonasschnelli commented:
Added memory dos prevention by filtering/only allowing valid commands.

sdaftuar commented on src/net.cpp:
```diff
@@ -667,6 +676,19 @@ bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes)
+            //store received bytes per command
+            //to prevent a memory DOS, only allow valid commands
+            std::string command(msg.hdr.pchCommand);
+            for (unsigned int i = 0; i < sizeof(logAllowIncommingCmds)/sizeof(logAllowIncommingCmds[0]); i++)
```
Since this is a map lookup, can we just initialize the map once with entries for each valid string, and then do a find() here?

jonasschnelli replied:
Yes. Would be possible. I somehow searched for a solution that doesn't require initialization. Do you think the `map.find()` would be faster or just less code?

sdaftuar replied:
Mostly I was thinking less code/easier to read (though it would be faster too, since we have to do a find() anyway).

laanwj replied:
C++11 to the rescue:
```c++
const static std::set<std::string> logAllowIncommingCmds = {
    "version", "verack", "addr", "inv", "getdata",
    "getblocks", "getheaders", "tx", "headers", "block",
    "getaddr", "mempool", "ping", "pong", "alert",
    "filterload", "filteradd", "filterclear", "reject"
};
```

```
    if (logAllowIncomingCmds.count(command))
        mapRecvBytesPerCmd[command] += msg.hdr.nMessageSize;
```

jgarzik commented:
ut ACK - this actually recreates a change I wrote a long time ago

dcousens commented:
concept ACK

jonasschnelli commented:
IMO
`for (unsigned int i = 0; i < sizeof(logAllowIncommingCmds)/sizeof(logAllowIncommingCmds[0]); i++)` is fine until we switch over to C++11.

@dcousens: not sure if disk persistence would make sense here. I'd like to extend this up to the GUI layer to have a better network/bandwidth graph that could also generate and show data when it's not opened, though, this PR does not include time based data logging.

jonasschnelli commented:
Passed Travis now.
Would be nice if someone can test this.

laanwj commented:
Going to test this

laanwj commented on src/net.cpp:
```diff
@@ -682,6 +691,19 @@ bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes)
+            //store received bytes per command
+            //to prevent a memory DOS, only allow valid commands
+            std::string command(msg.hdr.pchCommand);
+            for (unsigned int i = 0; i < sizeof(logAllowIncomingCmds)/sizeof(logAllowIncomingCmds[0]); i++)
```
I've thought of - what I think is an even better solution - as it doesn't require recourse to anything besides the map itself: prepopulate the `mapRecvBytesPerCmd`. At initialization:
```
    for (unsigned int i = 0; i < sizeof(logAllowIncomingCmds)/sizeof(logAllowIncomingCmds[0]); i++)
        mapRecvBytesPerCmd[logAllowIncomingCmds[i]] = 0;
```
Then only add if the item exists, e.g.
```c++
    mapCmdSize_t::iterator i = mapRecvBytesPerCmd.find(msg.hdr.pchCommand);
    if (i != mapRecvBytesPerCmd.end())
        i->second += 1;
```

laanwj commented on src/net.h:
```diff
@@ -199,7 +200,9 @@ class CNodeStats
     bool fInbound;
     int nStartingHeight;
     uint64_t nSendBytes;
+    mapCmdSize_t mapSendBytesPerCmd;
```
Nit: don't use `_t` for the type name.
- I've learned recently (think I heard it from @gmaxwell) that this is reserved by the C standard for system types
- We don't use this convention for any of our own types

jonasschnelli commented:
@laanwj: I like your map.find approach.
Implemented (squash-me commit) and renamed POSIX reserved `*_t` type.

laanwj commented on src/rpcnet.cpp:
```diff
@@ -165,6 +173,18 @@ UniValue getpeerinfo(const UniValue& params, bool fHelp)
+
+        UniValue recvPerCmd(UniValue::VOBJ);
+        BOOST_FOREACH( const mapCmdSize::value_type &i, stats.mapRecvBytesPerCmd) {
+            recvPerCmd.push_back(Pair(i.first, i.second));
```
If we want we could hide the entries with `i.second==0` here.

jonasschnelli replied:
I also though about that, but I think the null entries are better instead of hiding. It result in more controllable parsing.

laanwj replied:
Why would this result in more controllable parsing? The zeros don't mean anything. The only reason those entries are there in the first place is our 'ignore unknown command' implementation detail.

laanwj replied:
Otherwise, to be consistent, if you really want an overview of all possible messages, we want zero entries for `mapSendBytesPerCmd` too.

jonasschnelli replied:
I though if someone writes a statistics fronted that uses the RPC as data-source it would complicate things (a little bit) because you don't have a full list of available commands at the beginning. But no strong opinion. I think the drawback right now is a relatively big response.

laanwj replied:
OK i agree. If you want to know eg. the set of commands that is not considered *other*. OK I don't care deeply about this, fine to leave it as-is, but I expect more people to comment on this.

laanwj commented:
Minimally-tested ACK, I like this functionality a lot

laanwj commented:
Speaking of which, why is `verack` accounted for 0 bytes? I guess we don't include the message header in the size?
```
    "bytessent_per_cmd": {
      "addr": 31,
      "getheaders": 997,
      "inv": 9043,
      "ping": 16,
      "verack": 0,
      "version": 115
    },
```

jonasschnelli commented:
Correct. Right now only the message size counts... so your right. This needs to be fixed, otherwise we are not really counting net bytes. Working on it.

laanwj commented:
It's a fixed number per packet, so should be straightforward enough :)

jonasschnelli commented:
Added the header size to the bytes addition. Squashed to one commit.

jonasschnelli commented:
Implemented @laanwj  idea. Added a category `"*unknown*"` to aggregate bytes of received command that are not known. This could help identifying misbehaving nodes.

laanwj commented on src/net.cpp:
```diff
@@ -682,6 +684,15 @@ bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes)
+            //store received bytes per command
+            //to prevent a memory DOS, only allow valid commands
+            mapCmdSize::iterator i = mapRecvBytesPerCmd.find(msg.hdr.pchCommand);
+            if (i != mapRecvBytesPerCmd.end())
```
or
```
mapCmdSize::iterator i = mapRecvBytesPerCmd.find(msg.hdr.pchCommand);
if (i == mapRecvBytesPerCmd.end())
    i = mapRecvBytesPerCmd.find("*other*");
i->second += msg.hdr.nMessageSize + CMessageHeader::HEADER_SIZE;
```
:-)
(make sure `*other*` is prepopulated)

jonasschnelli commented:
Added a commit that hides 0 entries in `getpeerinfo`.

laanwj commented:
@jonasschnelli This currently runs into an assert, as mapRecvBytesPerCmd[NET_COMMAND_OTHER] doesn't exist

jonasschnelli commented:
> @jonasschnelli This currently runs into an assert, as mapRecvBytesPerCmd[NET_COMMAND_OTHER] doesn't exist

Just fixed this over a force push. I forgot to initialize the map entry.

laanwj commented on src/net.cpp:
```diff
@@ -67,6 +67,15 @@ namespace {
+    "version", "verack", "addr", "inv", "getdata",
+    "getblocks", "getheaders", "tx", "headers", "block",
+    "getaddr", "mempool", "ping", "pong", "alert",
+    "filterload", "filteradd", "filterclear", "reject"};
```
I noticed some `*other*` in the stats from normal Bitcoin Core client, so I checked all uses of PushMessage:
These need to be added: `, "merkleblock", "notfound"`

laanwj commented:
Small conceptual nit: In contrast to client-server protocols in a P2P protocol there are no 'commands', just 'messages'.
The ddistinction seems only a technicality, but some confusion stems from this - for example when people propose new add messages which have a mandatory response, thinking of it as a command/response protocol. 
I know we get this naming wrong in some places in the source code too so it's no big deal there... but e.g. in the UI we may want to get this right :-)

laanwj commented:
This works great though.
ACK after adding the two missing message types.

jonasschnelli commented:
Pushed to commits:
-> Add "merkleblock" and "notfound" to the list of accepted commands for logging.
-> Reword the variables, comments and help-text to reflect "message command"'s instead of just just "command".

paveljanik commented:
@jonasschnelli To be consistent, please also change `recvPerCmd`.

jonasschnelli commented:
@paveljanik: Thanks of the review! Change and force pushed the last commit.

paveljanik commented:
ACK
This is very useful, thanks for it!

MarcoFalke commented on src/rpcnet.cpp:
```diff
@@ -111,6 +111,14 @@ UniValue getpeerinfo(const UniValue& params, bool fHelp)
+            "       ...\n"
+            "    }\n"
+            "    \"bytesrecv_per_msg\": {\n"
+            "       \"addr\": n,             (numeric) The total bytes received aggregated by message command\n"
```
Nit: Not sure if "message command" is a proper word? What about "message type" or even just "message"?

laanwj replied:
My vote goes for "Message type" too

jonasschnelli replied:
Agreed. Changed to "message type" (amend force push last commit).

GIJensen commented:
ACK

GIJensen commented on src/rpcnet.cpp:
```diff
@@ -165,6 +173,20 @@ UniValue getpeerinfo(const UniValue& params, bool fHelp)
+        obj.push_back(Pair("bytessent_per_msg", sendPerMsgCmd));
+
+        UniValue recvPerCmd(UniValue::VOBJ);
+        BOOST_FOREACH( const mapMsgCmdSize::value_type &i, stats.mapRecvBytesPerMsgCmd) {
```
nit: I believe these should read "BOOST_FOREACH(const mapMsgdSize..." (No space between "(" and "const")

paveljanik commented on src/net.h:
```diff
@@ -182,6 +182,7 @@ struct LocalServiceInfo {
 
 extern CCriticalSection cs_mapLocalHost;
 extern std::map<CNetAddr, LocalServiceInfo> mapLocalHost;
+typedef std::map<std::string, uint64_t > mapMsgCmdSize; //command, total bytes
```
Space after `uint64_t`

jonasschnelli commented:
Fixed minor code style issues (three unnecessary spaces).

laanwj commented:
Will merge after squash and travis OK

jonasschnelli commented:
Squashed.

rebroad commented on src/net.cpp:
```diff
@@ -2480,6 +2503,9 @@ void CNode::EndMessage() UNLOCK_FUNCTION(cs_vSend)
     WriteLE32((uint8_t*)&ssSend[CMessageHeader::MESSAGE_SIZE_OFFSET], nSize);
 
+    //log total amount of bytes per command
+    mapSendBytesPerMsgCmd[std::string(pszCommand)] += nSize + CMessageHeader::HEADER_SIZE;
```
why not put ssSend.size() into a variable and use that?

rebroad commented:
Is anyone using this functionality? If so, can uses be documented somewhere?

MarcoFalke commented:
@rebroad I think the goal was to find out what msg consumes the most bytes.
